docs/rfcs/039-single-click-pg-upgrade.md

```diff
@@ -0,0 +1,199 @@
+set it in all the right places and make the switch atomic.
+
+#### Let's set the terminology:
+```
+
+v14 branch ---------------------|*****X
+                                |
+v15 initdb + v15 pg_restore  I====R
+                                |
+v15 branch                      |~~~~~S--------->
+
+```
+
+`|` - branch_lsn. We don't really use it anywhere in the pg_upgrade branching, just use as a reference point.
+
+`I` - initdb_lsn of a new branch. We don't use it anywhere.
+
+`R` - LSN of a new version branch pg_restore catalog import. Let's call it the `restore_lsn`.
+We don't care about the history before restore_lsn, becaues we import timeline into branch as a snapshot.
+The restore_lsn may be both smaller and larger than the branch_lsn, depending on how active the old branch was and how many objects were created in the new branch by pg_restore (width of `====` on picture). 
+
+`S` - `start_lsn` of a new version branch. 
+We upload the prepared snapshot of the new version branch to the storage as of lsn `S`.
+This is the similar mechanism that we use to import new main timeline into the storage, but here we skip all non-catalog relation files - new branch will read their content from the old branch.
+
+`****` - writes that happened in v14 during the upgrade process. v15 must be able to read them. 
+
+`X` - final lsn of the old branch. It must be smaller than S. We must somehow restrict writes to the old branch after the version switch. Otherwise we will have data between S and X that will not be readable from the new branch.
+
+#### Complicated steps:
+
+##### How to figure out the value of `S`:
+- It must be larger than branch_lsn `|` to kepp the history line correct.
+- It must be larger than `R` to ensure that we can read the new objects created by pg_restore. Because otherwise the pages of the new branch will contain LSNs from the future and XLOG flush will break badly.
+- It must be larger than `X` to ensure that we don't miss any changes that happened in the old branch during the upgrade.
+
+- IIUC the correct value is `S = MAXALIGN(X+1)`, because we cannot import new timeline at LSN that exists in parent. 
+**Need review from the storage team here** to understand how hard this restriction is and if there is any problem with +1.
+
+##### Where to set `S` for the new branch:
+
+- We need to use it as a start_lsn for the new Timeline struct
+- We need to set it in the control file of the new branch.
+- We need to set in in the Checkpoint record of the new branch.
+
+##### How to make the switch atomic:
+
+- We need to restrict writes to the old branch to get the `X` value and use it as `S` for the new branch.
+- IIUC, we need to do this right before the import. This means that timeline import time is customer visible pg_upgrade downtime. Can we somehow make it shorter? What is our target downtime for the upgrade? **Need storage team help here** 
```

Comment:
I am unclear what help is needed here. Isn't the only problem that in our prototype we were thinking of only using the branch ancestor_lsn for selecting the pg version, not another lsn which would allow for these `pg_upgrade` Lsns in the created branch?

We also want to deny branching from these `pg_upgrade` Lsns, which in my above proposal would be `ancestor_lsn...pg_upgraded_lsn` and the `pg_upgraded_lsn` would be the first branchable (as in, to get a new branch with the latest version data). This `pg_upgraded_lsn` corresponds to your `S`, I think?

Reply:
> Can we somehow make it shorter? What is our target downtime for the upgrade?

Let's make it work first, and try to optimize later. Downtime is OK for the first version, let's not get fixated on minimizing it.